### PR TITLE
support max decimal extension

### LINKED ISSUE #710: Element extension: maxDecimalPlaces
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Would you like to work on the issue?**
Please state if this issue should be assigned to you or who you think could help to solve this issue.

### PR DESCRIPTION
**Pull request by epicadk**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #710

**Description**
Clear and concise code change description. 

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?
We could add this to the Primitive max length validator. 

**Type**
Choose one: Feature

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

**Comment by epicadk** [comment]
This is my first pr on the data capture module. Could someone review it and tell me if something is missing?

**Comment by Tarun-Bhardwaj** [comment]
> This is my first pr on the data capture module. Could someone review it and tell me if something is missing?

@aditya-07 , could you please take a look at it?

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt [hunk 1]
```diff
@@ -34,16 +34,15 @@ import org.robolectric.annotation.Config
 @RunWith(RobolectricTestRunner::class)
 @Config(sdk = [Build.VERSION_CODES.P])
 class DecimalTypeMaxDecimalValidatorTest {
-
   lateinit var context: Context
 
   @Before
-  fun initContext() {
+  fun setupTest() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt:40]: conventionally we call this `setUp`

```suggestion
  fun setUp() {
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt [hunk 2]
```diff
@@ -107,10 +106,12 @@ class DecimalTypeMaxDecimalValidatorTest {
         context
       )
 
-    Truth.assertThat(validationResult.isValid).isFalse()
-    Truth.assertThat(validationResult.message)
-      .isEqualTo("The maximum number of decimal places that are permitted in the answer is: 2")
+    assertThat(validationResult.isValid).isFalse()
+    assertThat(validationResult.message)
+      .isEqualTo("The maximum number of decimal places that are permitted in the answer is:2")
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt:111]: hey i liked the space after the `:` :) can we add it back? thanks

File: datacapture/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -5,6 +5,11 @@
     <string
         name="min_value_validation_error_msg"
     >Minimum value allowed is:<xliff:g id="min_value">%1$s</xliff:g> </string>
+    <string
+        name="max_decimal_validation_error_msg"
+    >The maximum number of decimal places that are permitted in the answer is:<xliff:g
```
Comment (jingtang10) [inline datacapture/src/main/res/values/strings.xml:10]: add a space after `:`

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidator.kt [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import org.hl7.fhir.r4.model.IntegerType
+import org.hl7.fhir.r4.model.QuestionnaireResponse
+
+internal object DecimalTypeMaxDecimalValidator :
+  ValueConstraintExtensionValidator(
+    url = MAX_DECIMAL_URL,
+    predicate = {
+      extension: Extension,
+      answer: QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent ->
+      val maxDecimalValue = (extension.value as? IntegerType)?.value
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidator.kt:38]:
```suggestion
      val maxDecimalPlaces = (extension.value as? IntegerType)?.value
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt (hunk 1 above)
Comment (jingtang10) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt:40]: hm i realised this is the case for the other validator tests.... if you don't mind renaming all of them to `setUp` it'd be good. thanks!

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        QuestionnaireResponse.QuestionnaireResponseItemComponent(),
+        context
+      )
+
+    assertThat(validationResult.isValid).isTrue()
+    assertThat(validationResult.message.isNullOrBlank()).isTrue()
+  }
+
+  @Test
+  fun validAnswer_shouldReturnValidResult() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt:60]:
```suggestion
  fun validate_validAnswer_shouldReturnValidResult() {
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent()
+              .setValue(DecimalType("1.00"))
+          ),
+        context
+      )
+
+    assertThat(validationResult.isValid).isTrue()
+    assertThat(validationResult.message.isNullOrBlank()).isTrue()
+  }
+  @Test
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt:77]: add blank line

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  fun noExtension_shouldReturnValidResult() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt:78]:
```suggestion
  fun validate_noExtension_shouldReturnValidResult() {
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+          ),
+        context
+      )
+
+    assertThat(validationResult.isValid).isTrue()
+    assertThat(validationResult.message.isNullOrBlank()).isTrue()
+  }
+
+  @Test
+  fun invalidAnswer_shouldReturnInvalidResult() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt:95]:
```suggestion
  fun validate_tooManyDecimalPlaces_shouldReturnInvalidResult() {
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt (hunk 7 above)
Comment (jingtang10) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidatorTest.kt:78]: i would also move this test case up to the top (because it's kinda like a zero case)

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidator.kt [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+import org.hl7.fhir.r4.model.Questionnaire
+import org.hl7.fhir.r4.model.QuestionnaireResponse
+
+internal object DecimalTypeMaxDecimalValidator : ConstraintValidator {
+  override fun validate(
+    questionnaireItem: Questionnaire.QuestionnaireItemComponent,
+    questionnaireResponseItem: QuestionnaireResponse.QuestionnaireResponseItemComponent,
+    context: Context
+  ): ConstraintValidator.ConstraintValidationResult {
+
```
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidator.kt:39]: this wasn't done?

**Review by jingtang10 — APPROVED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidator.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (epicadk) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/validation/DecimalTypeMaxDecimalValidator.kt:39]: I rewrote the object by extending `ValueConstraintExtensionValidator` this seems to be before that
